Usage messages: deduplicate '(default true)' et al

edsantiago opened the pull request:
Remove hardcoded '(default: true)' strings from bool flags,
and '(default this-or-that)' from string flags.

First because it's unmaintainable duplication that would cause
confusion should someone ever change the default and not notice
the message.

Second, because cobra[1] already prints '(default XXXX)' for
all options with non-false non-nil default. So in each of
these cases, current podman help behavior is:

    $ podman login --help
    ...
       --tls-verify  Require HTTPS ... (default: true) (default true)

This PR eliminates that duplication.

 [1] actually spf13/pflag/flag.go

The only nontrivial one of these is start.go, where the default
for sigProxy depends on the --attach flag. Solution: change
the command-line default to false, and implement the new
conditional default in logic. Bonus: removed unnecessary
check, because now if sigProxy is set without --attach,
we can guarantee that it was done by the user. But please
pay close scrutiny to this particular section in case
there's something I missed.

Signed-off-by: Ed Santiago <santiago@redhat.com>

mheon commented on cmd/podman/common.go:
```diff
@@ -374,7 +374,7 @@ func getCreateFlags(c *cliconfig.PodmanCommand) {
 	createFlags.Int64(
 		"memory-swappiness", -1,
-		"Tune container memory swappiness (0 to 100) (default -1)",
+		"Tune container memory swappiness (0 to 100)",
```
This says 0 to 100 and then has a default of -1 and that does not seem right.

Does -1 represent no limit?

TomSweeneyRedHat replied:
-1 == take the default setting from the system.

mheon commented on cmd/podman/export.go:
```diff
@@ -36,7 +36,7 @@ func init() {
 	exportCommand.SetUsageTemplate(UsageTemplate())
 	flags := exportCommand.Flags()
-	flags.StringVarP(&exportCommand.Output, "output", "o", "/dev/stdout", "Write to a file, default is STDOUT")
+	flags.StringVarP(&exportCommand.Output, "output", "o", "/dev/stdout", "Write to a file")
```
For this and other STDOUT/STDIN things - should we say "instead of the terminal" to make it clear the default is the terminal?

edsantiago replied:
I think that makes it clearer, and am happy to change it, but would like confirming (or negating) opinions

edsantiago replied:
The key problem here is we have to deal with Cobra's automatic addition of "(default)":
```
  -o, --output string   Write output to FILE instead of stdout (default "/dev/stdout")
```
(And no, I can't figure out any way to tell Cobra to say "--output FILE" instead of "--output string"; but I haven't looked closely enough).

The only way to get rid of Cobra's "(default XXX)" is to make the default null. As you can see from the discussion in `start.go` (with `--sig-proxy`), that results in great uncertainty. I'm willing to try it, but before doing so, is there any way we can come up with a simple help message that looks good with the Cobra default?
```
  -o, --output string   INSERT HELPFUL MESSAGE HERE (default "/dev/stdout")
```
(IOW: we have control over HELPFUL MESSAGE; less so about the appended default).

mheon commented on cmd/podman/start.go:
```diff
@@ -62,14 +62,10 @@ func startCmd(c *cliconfig.StartValues) error {
 	}
 
-	sigProxy := c.SigProxy
+	sigProxy := c.SigProxy || attach
```
The old version had the advantage of detecting is `--sig-proxy` was explicitly set when `--attach` was not set and making that an error

edsantiago replied:
This preserves that behavior: see below. (The `.Changed` check is no longer necessary, because we are now guaranteed that the condition in line 67 can only happen if user explicitly sets `--sig-proxy` without `--attach`).

mheon replied:
Why the default change from false to true for sig-proxy?

edsantiago replied:
My intention is to preserve the existing behavior, as both documented and implemented. I believe that my changes preserve everything exactly as before. The only difference is that the Cobra default is now false instead of true; the purpose of that is to change the autogenerated --help message:
```
   old: --sig-proxy    Proxy received signals to the process (default true if attaching, false otherwise) (default true)
   new: --sig-proxy    Proxy received signals to the process (default true if attaching, false otherwise)
```
That is, remove the confusing duplicate '(default true)' from the usage message. AFAICT there is no way to tell Cobra "I've got the `default` message covered, please shut up"

mheon replied:
Oh oh oh... I think I see what's going on here. Sorry it took a minute, I need more coffee.

Yeah, this works fine.

TomSweeneyRedHat replied:
If @mheon's hip, I'm hip.  Too many reviews today.

TomSweeneyRedHat commented on cmd/podman/start.go:
```diff
@@ -41,7 +41,7 @@ func init() {
 	flags.BoolVarP(&startCommand.Interactive, "interactive", "i", false, "Keep STDIN open even if not attached")
 	flags.BoolVarP(&startCommand.Latest, "latest", "l", false, "Act on the latest container podman is aware of")
-	flags.BoolVar(&startCommand.SigProxy, "sig-proxy", true, "Proxy received signals to the process (default true if attaching, false otherwise)")
+	flags.BoolVar(&startCommand.SigProxy, "sig-proxy", false, "Proxy received signals to the process (default true if attaching, false otherwise)")
```
I don't think this should change here.

edsantiago replied:
@TomSweeneyRedHat could I ask you to take a step back, review the new logic, read the thread with @mheon, and then decide if you really mean that or not?

rhatdan commented:
LGTM other then the question about the -1

edsantiago commented:
Thank you all for helpful comments.

New force push: rebased, included @mheon's "instead of terminal" suggestion, and, as a non-squashed separate commit, a risky (IMO) change to the Cobra default for `memory-swappiness`: it is now `0` (zero). This makes Cobra not output the confusing '(default -1)' message in the --help. I believe I've found and fixed the only instance in the code base that references this command-line option, but I really have no idea what I'm doing so please feel free to tell me to back out this change.

mheon commented:
I think we ought to leave the -1, but figure out why it's set to -1, and note in the description text that -1 is a valid value (to use the system default, or unlimited, or whatever it happens to mean)

edsantiago commented:
@mheon, I agree. My previous change wasn't even working - it broke `podman run` on rootless. I've undone it, and simply changed the help message text:
```
  --memory-swappiness int    Tune container memory swappiness (0 to 100, or -1 for system default) (default -1)
```

rhatdan commented:
:+1:

TomSweeneyRedHat commented on cmd/podman/export.go:
```diff
@@ -36,7 +36,7 @@ func init() {
 	exportCommand.SetUsageTemplate(UsageTemplate())
 	flags := exportCommand.Flags()
-	flags.StringVarP(&exportCommand.Output, "output", "o", "/dev/stdout", "Write to a file, default is STDOUT")
+	flags.StringVarP(&exportCommand.Output, "output", "o", "/dev/stdout", "Write to a file instead of terminal")
```
sorry @edsantiago , I think this should be "instead of the terminal".  IDK I caught that earlier.  WDYT?

edsantiago replied:
@TomSweeneyRedHat this is under discussion now, see @rhatdan's [comment](https://github.com/containers/libpod/pull/2633#discussion_r265337021) about "terminal" actually being incorrect. You are good with words, I'd love to hear your suggestion about improved wording (or, if necessary, removing the Cobra default and initializing in code).

TomSweeneyRedHat replied:
I'm not a big fan of terminal, when I hear "terminal" I think VT100 Terminal (the real one, not the emulator).  Then afterwards I think of a linux software terminal, so I understand the hesitancy to use the term.

But I think it conveys the fact that the data is going to the screen attached to the process.  Saying STDOUT is closer to what we're doing, but we're also writing info to STDERR.

So I'd stick with "the terminal", but could easily be swayed as I've struggled describing similar circumstances.

TomSweeneyRedHat commented on cmd/podman/load.go:
```diff
@@ -34,7 +34,7 @@ func init() {
 	loadCommand.SetUsageTemplate(UsageTemplate())
 	flags := loadCommand.Flags()
-	flags.StringVarP(&loadCommand.Input, "input", "i", "/dev/stdin", "Read from archive file, default is STDIN")
+	flags.StringVarP(&loadCommand.Input, "input", "i", "/dev/stdin", "Read from archive file instead of from terminal")
```
If the added above before terminal, here too.

TomSweeneyRedHat commented on cmd/podman/save.go:
```diff
@@ -58,7 +58,7 @@ func init() {
 	flags.BoolVar(&saveCommand.Compress, "compress", false, "Compress tarball image layers when saving to a directory using the 'dir' transport. (default is same compression type as source)")
 	flags.StringVar(&saveCommand.Format, "format", v2s2Archive, "Save image to oci-archive, oci-dir (directory with oci manifest type), docker-archive, docker-dir (directory with v2s2 manifest type)")
-	flags.StringVarP(&saveCommand.Output, "output", "o", "/dev/stdout", "Write to a file, default is STDOUT")
+	flags.StringVarP(&saveCommand.Output, "output", "o", "/dev/stdout", "Write to a file instead of terminal")
```
If the added above before terminal, here too.

mheon commented:
/approve

edsantiago commented:
All right -- how's [this](https://github.com/containers/libpod/pull/2633/commits/1e124306dbd35a4cfdf3f585119a2c4441ec543d)? I've cleared the Cobra default in all three offending cases, improved (I believe) the usage message, and (fingers crossed) preserved existing behavior. Is this better?

And, hey, what's the PATH (imageName) argument for in `load`? It doesn't seem to be treated as a path, nor does it actually seem to have any effect whatsoever.

edsantiago commented:
/retest

TomSweeneyRedHat commented:
That wording works for me.  Tests are still complaining...

edsantiago commented:
Tests are failing all over the place; even a one-line string change (#2657). I don't think it's anything in my code.

rhatdan commented:
bot, retest this please

edsantiago commented:
Thanks, that did it.

Why is the magic word sometimes slash-re-test and sometimes the 'please' sentence?

rhatdan commented:
/lgtm

mheon commented:
@edsantiago We have three separate CI systems.

There's the legacy Papr stuff (the "Containerized (Podman in Podman)" and "Containerized (Podman in Docker)" tests), which is `bot, retest this please`

The Openshift tests (`ci/prow` tests), which are `/retest`

And the Cirrus tests, which you have to click on the "Checks" tab of the PR and then manually click the re-run button

Keeping track of this is not always easy
